bpfd-operator/controllers/bpfd-agent.go

```diff
@@ -0,0 +1,504 @@
+}
+
+const bpfdAgentFinalizer = "bpfd.io.agent/finalizer"
+const retryDuration = 10 * time.Second
+
+//+kubebuilder:rbac:groups=bpfd.io,resources=ebpfprograms,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=bpfd.io,resources=ebpfprograms/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=bpfd.io,resources=ebpfprograms/finalizers,verbs=update
+//+kubebuilder:rbac:groups=bpfd.io,resources=ebpfprogramconfigs,verbs=get;list;watch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the EbpfProgram object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+// This should be called in the following scenarios
+// 1. A new ebpfProgramConfig Object is created
+// 2. An ebpfProgramConfig Object is Updated (i.e one of the following fields change
+//   - NodeSelector
+//   - Priority
+//   - AttachPoint
+//   - Bytecodesource
+//
+// 3. Our NodeLabels are updated and the Node is no longer selected by an EbpfProgramConfig
+//
+// 4. And ebpfProgramCongfig Object is deleted
+//
+// To cover each case mentioned above the Reconcile loop does the following
+// 1. If (ebpfProgram object doesn't exist && our node is selected) Load and Create ebpfProgram object
+// 2. If (ebpfProgram exists && our node isn't selected) Remove and Delete ebpfProgram object
+// 3. If (ebpfProgram exists && our node is Selected)
+//   - List All programs
+//   - Get Priority, AttachPoint, and BytecodeSrc from output
+//   - Match ^ with what's in the current ebpfProgramConfig
+//   - Reconcile (Remove and Add) if information has changed
+//     -> Update program Map
+//
+// 4. If Deletion Timestamp is set on ebpfProgramConfig Remove and Delete ebpfProgram
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.13.0/pkg/reconcile
+func (r *EbpfProgramReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	l := log.FromContext(ctx)
+
+	l.Info("ebpf-agent is reconciling", "request", req.String())
+
+	// Lookup Ks node object for this bpfd-agent This should always succeed
+	ourNode := &v1.Node{}
+	if err := r.Get(ctx, types.NamespacedName{Namespace: v1.NamespaceAll, Name: r.NodeName}, ourNode); err != nil {
+		return ctrl.Result{Requeue: false}, fmt.Errorf("failed getting bpfd-agent node %s : %v",
+			req.NamespacedName, err)
+	}
+
+	ebpfProgramConfigs := &bpfdiov1alpha1.EbpfProgramConfigList{}
+
+	opts := []client.ListOption{}
+
+	if err := r.List(ctx, ebpfProgramConfigs, opts...); err != nil {
+		return ctrl.Result{Requeue: false}, fmt.Errorf("failed getting ebpfProgramConfigs for full reconcile %s : %v",
+			req.NamespacedName, err)
+	}
+
+	if len(ebpfProgramConfigs.Items) == 0 {
```

Comment:
Do we need a check to ensure no ebpfPrograms exist?

Reply:
No because there can/should not be any `ebpfProgram` Objects if there are no ebpfProgramConfig objects